segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,523 @@
+                            this.failedIteration.set(true);
+                        } else {
+                            this.failedIteration.set(false);
+                        }
+                        return null;
+                    });
+        }
+
+        /**
+         * Attempts to read data from the internal Segment. The size of the read length would be the minimum between the
+         * outstanding bytes and 2 * ProcessorEventData.MAX_TOTAL_EVENT_SIZE bytes.
+         *
+         * @return A {@link CompletableFuture} that, when completed, will contain a {@link EventsReadAndTruncationLength}
+         * object with events read and the next truncation length for the internal Segment.
+         */
+        private CompletableFuture<EventsReadAndTruncationLength> readEvents() {
+            return CompletableFuture.supplyAsync(() -> {
+                                        int readLength = (int) Math.min(getOutstandingBytes(), 2 * ProcessorEventData.MAX_EVENT_SIZE);
+                                        return this.segment.read(this.segment.getInfo().getStartOffset(), readLength, containerOperationTimeout);
+                                    }, this.executor)
+                                    .thenCompose(rr -> AsyncReadResultProcessor.processAll(rr, this.executor, containerOperationTimeout))
+                                    .thenApply(this::deserializeEvents);
+        }
+
+        @SneakyThrows(Exception.class)
+        private EventsReadAndTruncationLength deserializeEvents(BufferView inputData) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long truncationLength = 0;
+            Exception deserializationException = new NoDataAvailableException();
+            try {
+                @Cleanup
+                BoundedInputStream input = new BoundedInputStream(inputData.getReader(), inputData.getLength());
+                int dataLength = inputData.getLength();
+                while (input.getRemaining() > 0 && events.size() < this.config.getMaxItemsAtOnce()) {
+                    ProcessorEventData event = this.serializer.deserialize(input);
+                    events.add(event);
+                    // Update the truncation length, which includes user data and internal VersionedSerializer metadata.
+                    truncationLength = dataLength - input.getRemaining();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+            } catch (Exception ex) {
+                // Some unexpected serialization error occurred here; rethrow if this prevents reading further events.
+                log.error("{}: Problem while deserializing events.", this.traceObjectId);
```

Comment:
Added quite a bit of information in the `error` log when deserializing events which will provide good information in case we have issues related with data corruption or other serialization problems.